### PR TITLE
Generic GraphTypes, round 2

### PR DESCRIPTION
**Pull request by jquense**
A more holistic attempt at allowing generics here. I feel fairly happy about the structure of stuff, there is some duplication but it's not to bad.

The actual API here for declaring Fields is still pretty wonky. I did pull some of the Expression stuff in from `next`. I messed around with the methods there for a while but never ended up being very happy with it. You start to get ugly ambiguous overload methods between the Func and Expression versions. I'm also not a wiz with generics, so some of the FuncFieldResolver stuff is less than optimal (not leveraging TReturnType at all). Please let me know what you think :)

### CODE REVIEW

File: src/GraphQL/Types/ComplexGraphType.cs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        }
+
+        public FieldBuilder<TSourceType, TProperty> Field<TProperty>(
+            Expression<Func<TSourceType, TProperty>> expression,
+            bool nullable = false,
+            Type type = null)
+        {
+            string name;
+            try {
+                name = expression.NameOf().ToCamelCase();
```
Comment (joemcbride) [inline src/GraphQL/Types/ComplexGraphType.cs]: I am a fan of this (as I had it in the `next` branch), however I have heard from several people that they would like to have the option of PascalCase property names.  So this `ToCamelCase` will have to go to support that.

**Comment by joemcbride** [comment]
I'm liking this at first look. I would like to pull down the code and play with it a bit, but looking good. 👍

File: src/GraphQL/Types/ComplexGraphType.cs (hunk 1 above)
Comment (jquense) [inline src/GraphQL/Types/ComplexGraphType.cs]: does it make a difference in terms of final json payload? I went with camel case since it's more idiomatic json/graphql, but if that's handled by the json serializer sticking with pascal case here seems reasonable as well.

Comment (joemcbride) [inline src/GraphQL/Types/ComplexGraphType.cs]: It may make a difference, depending on what the final solution will be to enable the PascaleCase feature.  Right now it is all handled by the serializer and it is setup to just do camelCase.  Though to support a mix of both PascalCase and camelCase (since I believe some of the spec properties need to be camelCase) we may have to have it configured as "do not transform" or something.  I'm honestly not sure yet.

Comment (joemcbride) [inline src/GraphQL/Types/ComplexGraphType.cs]: What I'm kind of leaning towards is having the serializer setup as "do not transform", but then providing configurable convention.  So when a field is registered, even if someone mis-types the name, it would get transformed at registration time to that convention.  I'm leaning towards something like that to stay away from having JSON.NET attributes or something similar littered throughout the code.

Comment (jquense) [inline src/GraphQL/Types/ComplexGraphType.cs]: Ya it seems like you'd really want it to be a single sort of toggle for the entire schema vs annotating all the classes, since case is something you really want consistent. Like I said, as long as I can eventually consume the data in camelCase on the client it seems reasonable to not force something on the schema builder, while also letting them make use of shortcuts like deriving the name from the expression 👍

**Comment by jquense** [comment]
I do think the `Field` is not finished here, but I think the bones are there. I didn't want to presume changing all the API's without talking about it :P 

It seems like perhaps now that expressions are possible for infering field data, maybe the none-fluent form of the FuncResolver api is less relevant? Part of the problem is that that API tends to start clashing with the expression one. The other conflict is that the Func/Builder apis use generics to specify the GraphType, which is a heck of a lot neater than: `type: typeof(NonNullGraphType<ListGraphType<StringGraphType>>>)`. The nice bit about `next`s use of instances is that it deftly avoids that but not dealing in `Type`s.

One thought I had was to limit the non-expression API's to builders only, where you can more easily do `Type<StringGraphType>()`.

**Comment by jquense** [comment]
:wave: gonna rebase to fix conflicts.

I also was going to remove the `ToCamelCase` but that creates problems outside of just client json serialization. Field validation is also case sensitive and the schema fields will be in teh case you add them. 

I think for folks who want camelCase they want it in the schema in order to not fight the validation, or write queries in a different case then the payload actually is. To that end I imagine that you'd been the configuration option to be higher up than the json serializer, and I am not quite sure  how to go about that. It may be helpful to actually defer FieldType creation so that you can process the result

```csharp
Field(builder=> builder.Name("fOo"));
```
vs
```csharp
Field().Name("fOo"));
```

**Comment by joemcbride** [comment]
As you've seen, I haven't been able to get to this yet.

I'm okay with leaving the `ToCamelCase` for now, as other changes will be required for PascalCase to be fully supported.   I opened an issue on the graphql spec project to clarify their stance on what should be camelCase and what can be PascalCase and still be considered spec.  I say we can wait to do anything further to support PascalCase until I get a response.

I'll be honest, I don't like the fluent api as much as the other route of just parameters or a configuration object.  I can see how that is messy with dealing with types vs. instances though.

I want to move to instances eventually.  The major thing I wasn't sure how to solve initially was referencing a type that you may not have an instance of yet.  The Java project solved this by having a special "named reference" which gets resolved when the graph is initially built.

What I think the plan should be:

* get these changes in
* get the .NET Core changes in
* move to an instance-based API

**Comment by jquense** [comment]
my one sort of concern with instance based approach is that it makes defining stuff in seperate files much weirder. Seems like you'd end up defining classes anyway except now you need to call some sort of `Init()` method in the ctor. That may just be an artifact of the way `/next` is structured tho where the GraphType classes are more like builders and the `Config` is the real type data. 

That being said i've been trying to mess around with some Relay types and things like Mutations are way more convoluted and confusing using classes vs instances. Cases where you wnat to augment the graphType given to you (in this case adding `clientMutationId` and adjusting the resolvers to add it to the output type from the input).

Seems like it would be nice if you could mostly keep the current class structure, but you can jsut `new` them for use inside of defining fields. At this point tho i'm just talking a bunch about stuff that is outside my wheelhouse :)

**Comment by joemcbride** [comment]
I have to say I'm not as familiar with Relay in that regard.  I'll have to look at that more myself to know what is needed.

**Comment by jquense** [comment]
done rebasing! 

I'd also be happy to make an attempt at instance based things, but I'm a bit unsure how that changes the execution/building logic. i'm also fairly uncertain how to provide the right hooks for DI, ctor injections is the one i'm most familiar with.

**Comment by jquense** [comment]
related: it seems like you could handler recursive references via a closure, like the JS API.
```c#
ObjectGraphType<Droid> DroidType = null;

DroidType = new ObjectGraphType<Droid>(_ =>
{
    _.Name = "Droid";
    _.Description = "Droid type";
    _.Field(new ListGraphType(DroidType), "friends");
});
```

**Comment by joemcbride** [comment]
👍  That is kind of the beauty of the instance-based api.  Since the api will require instances, the responsibility is on the user to provide those instances.  The library doesn't have to deal with resolving them.

The way I saw the Java project solve the recursion issue is with something like this:


```
    public class GraphQLReference
    {
      public GraphQLReference(string name)
      {
        Name = name;
      }
      public string Name { get; }
    }

```

Then the framework replaces those references after the graph has been build.

**Comment by joemcbride** [comment]
```
var DroidType = new ObjectGraphType<Droid>(_ =>
{
    _.Name = "Droid";
    _.Description = "Droid type";
    _.Field(new ListGraphType(new GraphQLReference("Droid")), "friends");
});
```

**Comment by joemcbride** [comment]
This is what I added for the vNext branch:

```
/// <summary>
/// A special type to allow an object/interface to reference itself.
/// It is replaced with the real type object when the schema is built.
/// </summary>
public class GraphQLTypeReference : GraphQLType, IGraphQLOutputType
{
    public GraphQLTypeReference(string typeName)
    {
        Name = $"ref({typeName})";
        TypeName = typeName;
    }

    public string TypeName { get; set; }
}
```

**Comment by jquense** [comment]
interesting...does that approach offer something over using a closure?

**Comment by joemcbride** [comment]
I believe the main purpose would be to allow the creation of classes vs. inline instances.  Otherwise your schema has to be all in one place.

```csharp
public class DroidType : ObjectGraphType<Droid>
{
  public DroidType(/* inject dependencies here */)
  {
    Name = "Droid";
    Field(new ListGraphType(new GraphQLReferenceType(Name)), "friends");
  }
}
```

**Comment by joemcbride** [comment]
Well, maybe you can just use `this` there?

File: src/GraphQL/TypeExtensions.cs [hunk 2]
```diff
@@ -11,16 +12,68 @@ public static T As<T>(this object item)
...
+
+            if (type == typeof(string))
+            {
+                graphType = typeof(StringGraphType);
+            }
+
+            if (type == typeof(bool))
+            {
+                graphType = typeof(BooleanGraphType);
+            }
```
Comment (joemcbride) [inline src/GraphQL/TypeExtensions.cs]: Noticed that `DateTime` is missing in this check.  We have a `DateGraphType` that can be used.  We may eventually want to provide a hook so users can add their own custom Type->GraphType mapping here.

**Comment by jquense** [comment]
:tada: thanks!

**Comment by joemcbride** [comment]
@jquense Any objections to me pushing these changes combined with the .NET Core support?

**Comment by joemcbride** [comment]
@jquense Let me rephrase - would you like a different nuget than one that is also for .NET Core.  I'm going to push one for .NET Core that has these changes.

**Comment by jquense** [comment]
I'm actually not really familiar with the new Core stuff. Does the .net core only work with .net core or can I use it in projects on 4.5 (where i'm stuck for stuff)?

**Comment by joemcbride** [comment]
The nuget supports multiple frameworks.  The minimum would be 4.5.1 though.

https://www.nuget.org/packages/GraphQL/0.12.0-alpha-536

Once I merge the .NET Core code to master (which I want to do here pretty soon), you will also need to have .NET Core installed to develop the project.

**Comment by joemcbride** [comment]
https://www.nuget.org/packages/GraphQL/0.12.0-alpha-538

That one should work with base 4.5.

**Comment by randyridge** [comment]
It shouldn't be necessary to list each framework and net standard, just go with the lowest common denominator. Same applies for the parser project, it's netstandard is much too high and has like 87 framework versions :D

**Comment by joemcbride** [comment]
@randyridge Hah, good points.  Being a net standard newb I was just going with the defaults.  It looks like perhaps we can bump netstandard down to 1.1.

**Comment by randyridge** [comment]
Yeah, I believe that's the case, would need to doublecheck with the portability analyzer. Keep the library reference at 1.6, but declare the framework to the lowest.
